src/dune_engine/rule_cache.ml

```diff
@@ -177,18 +182,28 @@ module Workspace_local = struct
          we still re-run all the previous stages, which is a bit of a waste. We
          could remember what stage needs re-running and only re-run that (and
          later stages). *)
-      let rec loop stages =
+      let rec loop stages needed_deps =
         match stages with
-        | [] -> Fiber.return (Hit produced_targets)
+        | [] ->
+          if fst needed_deps = Dep.Set.empty
+          then Fiber.return (Hit produced_targets)
+          else
+            let open Fiber.O in
+            let deps, old_digest = needed_deps in
+            let* deps = Memo.run (build_deps deps) in
+            let new_digest = Dep.Facts.digest deps ~env in
+            (match Digest.equal old_digest new_digest with
+             | true -> Fiber.return (Hit produced_targets)
+             | false -> Fiber.return (Miss Miss_reason.Needed_deps_changed))
         | (deps, old_digest) :: rest ->
           let open Fiber.O in
           let* deps = Memo.run (build_deps deps) in
           let new_digest = Dep.Facts.digest deps ~env in
           (match Digest.equal old_digest new_digest with
-           | true -> loop rest
+           | true -> loop rest needed_deps
            | false -> Fiber.return (Miss Miss_reason.Dynamic_deps_changed))
       in
-      loop prev_trace.dynamic_deps_stages
+      loop prev_trace.dynamic_deps_stages prev_trace.needed_deps
```

Comment:
This is overly complicated; the check should have the form `check1 prev_trace.dynamic_deps_stages || check2 prev_trace.neede_deps`, there is no need to mix the check for both types of dependencies.